flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2682,15 +2682,15 @@ class ArrayExprLowering {
     llvm::SmallVector<mlir::Value> extents;
     for (auto extent : seqTy.getShape())
       extents.push_back(builder.createIntegerConstant(loc, idxTy, extent));
-    auto getArrayBox = [&]() {
-      if constexpr (A::category == Fortran::common::TypeCategory::Character) {
-        auto len = builder.createIntegerConstant(loc, idxTy, x.LEN());
-        return fir::CharArrayBoxValue{addr, len, extents};
-      } else {
-        return fir::ArrayBoxValue{addr, extents};
-      }
-    };
-    auto lambda = genarr(getArrayBox());
+    CC lambda;
+    if (fir::isa_char(seqTy.getEleTy())) {
+      auto charTy = seqTy.getEleTy().cast<fir::CharacterType>();
```

Comment:
Nit: if you need to cast it anyway, you could do:
```
if (auto charTy = seqTy.getEleTy().dyn_cast<fir::CharacterType>()) {
  ...
}
```

Reply:
Thanks, Jean.  I tried this unsuccessfully with `cast` instead of `dyn_cast`.  This one works.